src/runtime-rs/crates/hypervisor/src/device/device_manager.rs

```diff
@@ -0,0 +1,362 @@
+            self.unset_sandbox_block_index(index)?;
+            self.devices.remove(&id);
+            return Err(e);
+        }
+
+        Ok(id)
+    }
+
+    pub async fn try_remove_device(&mut self, device_id: &str, h: &dyn Hypervisor) -> Result<()> {
+        if let Some(dev) = self.devices.get(device_id) {
+            let skip = dev.lock().await.decrease_attach_count().await?;
+            if skip {
+                return Ok(());
+            }
+            if let Err(e) = dev.lock().await.detach(h).await {
+                dev.lock().await.increase_attach_count().await?;
+                return Err(e);
+            }
+            self.devices.remove(device_id);
+        } else {
+            return Err(anyhow!(
+                "device with specified ID hasn't been created. {}",
+                device_id
+            ));
+        }
+        Ok(())
+    }
+
+    pub async fn generate_agent_device(&self, device_id: String) -> Result<AgentDevice> {
+        // Safe because we just attached the device
+        let dev = self.get_device_by_id(&device_id).await.unwrap();
+        let base_info = dev.lock().await.get_device_info().await?;
+        let mut device = AgentDevice {
+            container_path: base_info.container_path.clone(),
+            ..Default::default()
+        };
+
+        match self.get_block_driver().await {
+            VIRTIO_MMIO => {
+                if let Some(path) = base_info.virt_path {
+                    device.id = device_id;
+                    device.field_type = KATA_MMIO_BLK_DEV_TYPE.to_string();
+                    device.vm_path = path;
+                }
+            }
+            VIRTIO_BLOCK => {
+                if let Some(path) = base_info.pci_addr {
+                    device.id = device_id;
+                    device.field_type = KATA_BLK_DEV_TYPE.to_string();
+                    device.vm_path = path;
+                }
+            }
+            _ => (),
+        }
+        Ok(device)
+    }
+
+    pub async fn get_block_driver(&self) -> &str {
+        self.block_driver.as_str()
+    }
+
+    pub async fn get_device_guest_path(&self, id: &str) -> Option<String> {
+        if let Some(device) = self.devices.get(id) {
+            if let Ok(dev_info) = device.lock().await.get_device_info().await {
+                return dev_info.virt_path;
+            }
+        }
+        None
+    }
+
+    async fn attach_device(
+        &mut self,
+        id: &str,
+        h: &dyn Hypervisor,
+        da: DeviceArgument,
+    ) -> Result<()> {
+        if let Some(dev) = self.devices.get(id) {
+            dev.lock().await.attach(h, da).await?;
+        } else {
+            return Err(anyhow!(
+                "device with specified ID hasn't been created. {}",
+                id
+            ));
+        }
+        Ok(())
+    }
+
+    async fn try_create_device(&mut self, dev_info: &mut GenericConfig) -> Result<ArcBoxDevice> {
+        if dev_info.major != 0 || dev_info.minor != 0 {
```

Comment:
0 is valid device number. Should we use the same condition in function `find_device`?

Reply:
fixed